packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -278,9 +278,20 @@ export class ConnectionManager implements IConnectionManager {
 
 	public shouldJoinWrite(): boolean {
 		// We don't have to wait for ack for topmost NoOps. So subtract those.
-		return (
-			this.clientSequenceNumberObserved < this.clientSequenceNumber - this.localOpsToIgnore
-		);
+		const outstandingOps =
+			this.clientSequenceNumberObserved < this.clientSequenceNumber - this.localOpsToIgnore;
+
+		// Previous behavior was to force write mode here only when there are outstanding ops (besides
+		// no-ops). The dirty signal from runtime should provide the same behavior, but also support
+		// stashed ops that weren't submitted to container layer yet. For safety, we want to retain the
+		// same behavior whenever dirty is false.
+		const isDirty = this.containerDirty();
+		if (outstandingOps !== isDirty) {
```

Comment:
I think the event should include outstanindOps, and isDirty, so we can evaluate the data with context. As i think it is expected that outstanding is false, and is dirty is true for stashed ops. I don't think we ever expect outstanding to be true, and is dirty to be false? or maybe only in the case of code proposal?